File: Problems/boj_나무자르기.py
#https://www.acmicpc.net/problem/2805
#Binary Search

import sys
input = sys.stdin.readline

n, m = map(int, input().split())
l = list(map(int, input().split()))

s = 0
e = int(1e9) #max(l) 쓰면 시간초과 났다.

result = 0
while s<=e:
    tree = 0
    mid = (s+e)//2
    for i in l:
        if i>mid:
            tree += i-mid

    if tree>=m:
        s = mid+1
        result = mid
    else:
        e = mid-1

print(result)

File: Problems/cpp/boj_나무자르기.cpp
//https://www.acmicpc.net/problem/2805
//Binary Search

#include <iostream>
#include <vector>

using namespace std;

int main() {
    ios::sync_with_stdio(0);
    cin.tie(0); cout.tie(0);

    int n, m, h;
    long long s=0, e=2e9;
    cin >> n >> m;

    vector<int> vec(n);
    for (int i=0; i<n; i++)
        cin >> vec[i];
    
    while (s<=e){
        long long mid = (s+e)/2;
        long long len = 0;

        for (const int v : vec){
            if (v>mid)
                len += v-mid;
        }

        if (len>=m){
            s = mid+1;
            h = mid;
        }
        else{
            e = mid-1;
        }
    }

    cout << h;
}